src/sgame/sg_bot_parse.cpp

```diff
@@ -148,9 +148,48 @@ static AIValue_t haveUpgrade( gentity_t *self, const AIValue_t *params )
 		&& BG_InventoryContainsUpgrade( upgrade, self->client->ps.stats ) );
 }
 
-static AIValue_t percentAmmo( gentity_t *self, const AIValue_t* )
+// This function returns the ratio of 2 ammo values (ammo/clips)
+// for a specific weapon.
+// Only used to reduce duplicated code.
+// TODO: make this a weapon_t or weaponAttributes_t method?
+static AIValue_t Ratio( int current, int max, bool infiniteAmmo )
 {
-	return AIBoxFloat( PercentAmmoRemaining( BG_PrimaryWeapon( self->client->ps.stats ), &self->client->ps ) );
+	if ( max == 0 || infiniteAmmo )
+	{
+		return AIBoxFloat( 1.f );
+	}
+	return AIBoxFloat( current / static_cast<float>( max ) );
+}
+
+// Return emptyness ammo ratio of currently loaded clip
+// TODO: implement for non-primary weapon (such as barbs or future secondary weapons)
+static AIValue_t percentAmmoClip( gentity_t *self, const AIValue_t* )
+{
+	ASSERT( self && self->client );
+	playerState_t const& ps = self->client->ps;
+	weaponAttributes_t const* wpa = BG_Weapon( BG_PrimaryWeapon( ps.stats ) );
+
+	if ( wpa->maxAmmo == 0 && !wpa->infiniteAmmo )
```

Comment:
The warning seems unnecessary as you already check these conditions in `Ratio` and handle them correctly. And it would print every frame here, so if you really want a warning it should be in the weapon parser code.

Reply:
Fair. I'll remove this. I guess I was just going paranoid here, or maybe it's simply older than the implementation of `Ratio`. I don't remember history details behind this.